### PR TITLE
Add more Java Api

### PR DESCRIPTION
**Pull request by 2m**
These were discovered while working on a play-ws Akka Http backend.

### CODE REVIEW

File: akka-http-core/src/main/java/akka/http/javadsl/model/ContentTypes.java [hunk 1]
```diff
@@ -28,6 +28,13 @@ private ContentTypes() { }
   public static final ContentType.WithCharset TEXT_CSV_UTF8 =
           akka.http.scaladsl.model.ContentTypes.text$divcsv$u0028UTF$minus8$u0029();
 
+  public static final ContentType.Binary NO_CONTENT_TYPE =
+          akka.http.scaladsl.model.ContentTypes.NoContentType();
```
Comment (jrudolph) [inline akka-http-core/src/main/java/akka/http/javadsl/model/ContentTypes.java]: Weird indentation but it seems to be consistent with the rest of the file ;)

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpEntity.scala [hunk 2]
```diff
@@ -172,6 +172,9 @@ sealed trait HttpEntity extends jm.HttpEntity {
   override def toStrict(timeoutMillis: Long, materializer: Materializer): CompletionStage[jm.HttpEntity.Strict] =
     toStrict(timeoutMillis.millis)(materializer).toJava
 
+  /** Java API */
+  override def withContentType(contentType: jm.ContentType): HttpEntity =
+    withContentType(contentType.asInstanceOf[ContentType])
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpEntity.scala]: We usually use `JavaMappings` and `.asScala`.

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala [hunk 3]
```diff
@@ -164,6 +166,14 @@ sealed trait HttpMessage extends jm.HttpMessage {
   }
   /** Java API */
   def addHeaders(headers: JIterable[jm.HttpHeader]): Self = mapHeaders(_ ++ headers.asScala.asInstanceOf[Iterable[HttpHeader]])
+  /** Java API */
+  def withHeaders(headers: JIterable[jm.HttpHeader]): Self =
+    withHeaders(headers.asScala.toIndexedSeq.asInstanceOf[immutable.Seq[HttpHeader]])
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala]: How about `.asScala.toVector`?

File: akka-http-core/src/main/java/akka/http/javadsl/model/Authority.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+   * Returns a Authority created by parsing the given string representation.
+   */
+  public static Authority create(String authority) {
+    return akka.http.scaladsl.model.Uri.Authority$.MODULE$.parse(
+      akka.parboiled2.ParserInput$.MODULE$.apply(authority),
+      Charset.forName("UTF8"),
+      akka.http.scaladsl.model.Uri$ParsingMode$Relaxed$.MODULE$
+    );
+  }
+}
```
Comment (ktoso) [inline akka-http-core/src/main/java/akka/http/javadsl/model/Authority.java]: hah weird that it was missing indeed

File: akka-http-core/src/main/java/akka/http/javadsl/model/HttpHeader.java [hunk 5]
```diff
@@ -44,4 +44,22 @@
...
+    public static HttpHeader parse(String name, String value) {
+      final akka.http.scaladsl.model.HttpHeader.ParsingResult result =
+        akka.http.scaladsl.model.HttpHeader.parse(name, value,
+          akka.http.impl.model.parser.HeaderParser$.MODULE$.DefaultSettings());
+
+      if (result instanceof akka.http.scaladsl.model.HttpHeader$ParsingResult$Ok) {
+        return ((akka.http.scaladsl.model.HttpHeader$ParsingResult$Ok) result).header();
+      }
+      else {
+        throw new IllegalArgumentException("Unable to parse header [" + name + "] with value [" + value + "]");
```
Comment (ktoso) [inline akka-http-core/src/main/java/akka/http/javadsl/model/HttpHeader.java]: I wonder if safe to include the value here hm... If it's some auth header etc?

**Comment by jrudolph** [comment]
Still need mima fixes.

File: akka-http-core/src/main/java/akka/http/javadsl/model/HttpHeader.java (hunk 5 above)
Comment (2m) [reply akka-http-core/src/main/java/akka/http/javadsl/model/HttpHeader.java]: Good point. Remove header name and value from the exception message.

**Comment by 2m** [comment]
PLS BUILD

**Comment by 2m** [comment]
Failure was https://github.com/akka/akka-http/issues/1685

**Comment by jrudolph** [comment]
PLS BUILD

**Comment by jrudolph** [comment]
PLS BUILD

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala [hunk 6]
```diff
@@ -164,6 +166,16 @@ sealed trait HttpMessage extends jm.HttpMessage {
   }
   /** Java API */
   def addHeaders(headers: JIterable[jm.HttpHeader]): Self = mapHeaders(_ ++ headers.asScala.asInstanceOf[Iterable[HttpHeader]])
+  /** Java API */
+  def withHeaders(headers: JIterable[jm.HttpHeader]): Self = {
+    import JavaMapping.Implicits._
+    withHeaders(headers.asScala.toVector.map(_.asScala))
+  }
+  /** Java API */
+  def toStrict(timeoutMillis: Long, ec: Executor, materializer: Materializer): CompletionStage[Self] = {
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala]: Good that it was as simple as that in the end :)

File: akka-http-core/src/main/java/akka/http/javadsl/model/HttpHeader.java [hunk 7]
```diff
@@ -44,4 +44,22 @@
...
+    public static HttpHeader parse(String name, String value) {
+      final akka.http.scaladsl.model.HttpHeader.ParsingResult result =
+        akka.http.scaladsl.model.HttpHeader.parse(name, value,
+          akka.http.impl.model.parser.HeaderParser$.MODULE$.DefaultSettings());
+
+      if (result instanceof akka.http.scaladsl.model.HttpHeader$ParsingResult$Ok) {
+        return ((akka.http.scaladsl.model.HttpHeader$ParsingResult$Ok) result).header();
+      }
+      else {
+        throw new IllegalArgumentException("Unable to parse header");
```
Comment (jrudolph) [inline akka-http-core/src/main/java/akka/http/javadsl/model/HttpHeader.java]: In the best case, we would also support the ParsingResult in some way on the Java side. If we don't want to do that we could at least throw an `IllegalHeaderException` or `ParsingError` with the `ErrorInfo` in the error. Then the user can decide whether to expose only the `summary` of the error info or the full information.

Comment (2m) [reply akka-http-core/src/main/java/akka/http/javadsl/model/HttpHeader.java]: Alright. Switched to `IllegalHeaderException`

Comment (jrudolph) [reply akka-http-core/src/main/java/akka/http/javadsl/model/HttpHeader.java]: Cool, thanks.

**Pull request merged**